core/src/cost_terms.cpp

```diff
@@ -0,0 +1,142 @@
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of Bielefeld University nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Authors: Michael Goerner */
+
+#include <moveit/task_constructor/cost_terms.h>
+#include <moveit/task_constructor/stage.h>
+
+#include <moveit/robot_trajectory/robot_trajectory.h>
+#include <moveit/planning_scene/planning_scene.h>
+
+#include <moveit/collision_detection/collision_common.h>
+
+#include <boost/format.hpp>
+
+namespace moveit {
+namespace task_constructor {
+namespace cost {
+
+double PathLength(const SubTrajectory& s) {
+	const auto& traj = s.trajectory();
+
+	if (traj == nullptr)
+		return 0.0;
+
+	double path_length{ 0.0 };
+	for (size_t i = 1; i < traj->getWayPointCount(); ++i)
+		path_length += traj->getWayPoint(i - 1).distance(traj->getWayPoint(i));
+	return path_length;
+}
+
+double TrajectoryDuration(const SubTrajectory& s) {
+	return s.trajectory() ? s.trajectory()->getDuration() : 0.0;
+}
+
+double LinkMotion::operator()(const SubTrajectory& s, std::string& comment) {
+	const auto& traj{ s.trajectory() };
+
+	if (traj == nullptr || traj->getWayPointCount() == 0)
+		return 0.0;
+
+	if (!traj->getWayPoint(0).knowsFrameTransform(link_name)) {
+		boost::format desc("LinkMotionCost: frame '%1%' unknown in trajectory");
+		desc % link_name;
+		comment = desc.str();
+	}
+
+	double distance{ 0.0 };
+	Eigen::Translation3d position{ traj->getWayPoint(0).getFrameTransform(link_name).translation() };
+	for (size_t i{ 1 }; i < traj->getWayPointCount(); ++i) {
+		Eigen::Translation3d new_position{ traj->getWayPoint(i).getFrameTransform(link_name).translation() };
+		distance += (new_position.vector() - position.vector()).norm();
+		position = new_position;
```

Comment:
nitpick (doesn't require copying values):
```suggestion
		std::swap(position, new_position);
```